File: Feelings_Site/secretsanta/scripts/secret_santa.py
import yaml
import random

from django.core.mail import send_mail
from django.conf import settings

def assign(people):
    """
    In: Dictionary with people an associated info
    Out: Different dictionary but that follows secret santa assignments, for one large loop
    """
    assignments = {}

    # Get names of all people participating (keys of people to name list)
    names = list(people.keys())
    firstName = names[0]


    # For each santa, assign a random person, repicking the random person until it is
    # not the same as the santa
    for santa in people:
        while True:
            giftee = random.choice(names)
            if santa == giftee:
                if len(names) == 1:
                    # We are in the case where there is only one giftee left and it's
                    # the same as santa. Exit and try again.
                    return None, True
            else:
                break


        assignments[santa] = giftee
        names.remove(giftee)

    # Check for short loops
    loopChecker = {}
    while firstName not in loopChecker:
        loopChecker[firstName] = True
        firstName = assignments[firstName]

    print(str(len(loopChecker)), str(len(assignments)))
    if len(loopChecker) != len(assignments):
        # There exists a loop otherwise everyone in assignments should be covered.
        return None, True


    return assignments, None


def email(santa_name, giftee_name, santa_email, giftee_address, giftee_hobbies):
    '''
    Sends an email based on santa and giftee information and assignment using django's
    email wrappers
    '''
    subject = "Feelings Secret Santa 2018"
    message = "Hello {},\n\nThank you for participating in this year's Feelings Secret Santa! Here is your assignment information:\n\n\tName: {}\n\tAddress: {}\n\tHobbies/Interests: {}\n\nThings to consider:\n- Please schedule your gift shipping so that it arrives as close to Christmas as possible, unless otherwise noted. Keep in mind that shipping times will generally be longer than normal during the holidays.\n- There is no strict limit to pricing, but $20-$30 is a good range to be in.\n- If you are having trouble deciding on a gift, feel free to ask around. Just be subtle, please.\n- Feel free to message me on Discord (dzy#0001) if you have any questions or concerns.\n\n\nSincerely,\nDnan".format(santa_name, giftee_name, giftee_address, giftee_hobbies)

    email_from = settings.EMAIL_HOST_USER
    recipient = [santa_email]

    send_mail(
        subject,
        message,
        email_from,
        recipient,
        fail_silently=False,
    )

    return True



# Read in YAML file without errors
# refer to people.yml for example

# change this to the real file before sending
with open("./secretsanta/scripts/people2.yml", "r") as stream:
    try:
        people = yaml.load(stream)
    except yaml.Error as exc:
        print(exc)

# make assignments for people (keys of a dict?)
while True:
    assignments, err = assign(people)
    if err is None:
        break

print(people)
print(assignments)

# send emails based on assignments
for santa, giftee in assignments.items():
    santa_name = people[santa]['name']
    giftee_name = people[giftee]['name']
    santa_email = people[santa]['email']
    giftee_address = people[giftee]['address']
    giftee_hobbies = people[giftee]['wishlist']

    email(santa_name, giftee_name, santa_email, giftee_address, giftee_hobbies)

File: Feelings_Site/secretsanta/scripts/secret_santa.go
package main

import (
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"math/rand"
	"net/smtp"
	"strings"

	yaml "gopkg.in/yaml.v2"

	"github.com/golang/glog"
)

type People struct {
	Name     string
	Email    string
	Address  string
	Wishlist string
}

type Email struct {
	SmtpServer      string `yaml:"smtp_server"`
	Domain          string
	AccountAddress  string `yaml:"account_address"`
	AccountPassword string `yaml:"account_password"`
	Port            int
}

var test = flag.Bool("test", true, "Pass false to this flag to run for realsies")

var emailTemp = `Subject: Feelings: 2016 Secret Santa

 Hello %s,
 
 Thank you for participating in ths year's Feelings Secret Santa! Here is your assignment information:

 Name: %s

 Address: %s

 Hobbies/interests: %s
 
 Things to consider:
 - Please schedule your gift shipping so that it arrives as close to Christmas as possible, unless otherwise noted. Keep in mind that shipping times will generally be longer than normal during the holidays.
 - There is no strict limit to pricing, but $20-$30 is a good range to be in.
 - If you are having trouble deciding on a gift, feel free to ask around. Just be subtle, please.
 - If you are STILL having trouble deciding on a gift, here is a Dan-approved list of gifts that everyone will love: https://amzn.com/B014P8L66E, https://amzn.com/B01DVJMI0Q, https://amzn.com/B002ECDZCW, https://amzn.com/B01H0AX66W 
 - Feel free to message me (Daniel) if you have any questions or concerns.
 
 Sincerely,
 Daniel`

func main() {
	flag.Parse()
	peopleConfig, err := ioutil.ReadFile("people.yml")
	if err != nil {
		glog.Fatalf(`error reading "people.yml" file: %s`, err)
	}
	people := map[string]*People{}
	err = yaml.Unmarshal(peopleConfig, people)
	if err != nil {
		glog.Fatalf(`error unmarshaling people config: %s`, err)
	}

	var assignments map[string]string
	for {
		assignments, err = assign(people)
		if err == nil {
			break
		}
	}
	glog.Infof("assignments are %+v", assignments)

	emailConfig, err := ioutil.ReadFile("smtp.yml")
	if err != nil {
		glog.Fatalf(`error reading "smtp.yml" file: %s`, err)
	}
	email := &Email{}
	err = yaml.Unmarshal(emailConfig, email)
	if err != nil {
		glog.Fatalf(`error unmarshaling email config: %s`, err)
	}

	auth := smtp.PlainAuth("", email.AccountAddress, email.AccountPassword, email.SmtpServer)

	for santa, giftee := range assignments {
		config := people[giftee]
		emailMsg := fmt.Sprintf(emailTemp, strings.Title(santa), strings.Title(config.Name), config.Address, config.Wishlist)
		glog.V(3).Infof("emailMsg for %s  with email %s is %s", santa, people[santa].Email, emailMsg)
		if !*test {
			err := smtp.SendMail(fmt.Sprintf("%s:%d", email.SmtpServer, email.Port),
				auth, email.AccountAddress, []string{people[santa].Email}, []byte(emailMsg))
			if err != nil {
				glog.Fatalf("Failed to send email: %s", err)
			}
		}
	}
}

func assign(people map[string]*People) (map[string]string, error) {
	assignments := make(map[string]string, len(people))

	// Get the names of all the people participating (keys to people)
	names := []string{}
	for name := range people {
		names = append(names, name)
	}
	firstName := names[0]

	// For each santa, assign a random person, repicking the random person until it is
	// not the same as the santa
	for santa := range people {
		var giftee string
		var i int
		for {
			i = rand.Intn(len(names))
			giftee = names[i]
			if santa == giftee {
				if len(names) == 1 {
					// We are in the case where there is only once choice of giftee left, and it
					// is the santa. Exit and try again.
					return nil, errors.New("Bad")
				}
			} else {
				break
			}
		}
		assignments[santa] = giftee
		names = append(names[:i], names[i+1:]...)
	}

	// Check for short loops
	loopChecker := make(map[string]bool)
	for !loopChecker[firstName] {
		loopChecker[firstName] = true
		firstName = assignments[firstName]
	}
	if len(loopChecker) != len(assignments) {
		// There is a loop somewhere, otherwise we should've covered
		// everyone in assignments
		return nil, errors.New("Bad")
	}

	return assignments, nil
}
